### PR TITLE
fix: pad input in policy inference

### PR DESCRIPTION
**Pull request by ypang-neuraco**
### Bugfixes
- For cross-embodiment training with 2 robots of different number of joints, the inference fails when using the robot with less joint because the input is not padded
- This fix pads the input data to the max items trained on similar to what we do in training
- Improve handling of padding in training
### Items
 - [NCRL-599](https://neuracore.atlassian.net/browse/NCRL-599)

### CODE REVIEW

File: neuracore/ml/utils/policy_inference.py [hunk 1]
```diff
@@ -142,34 +142,42 @@ def _preprocess(self, sync_point: SynchronizedPoint) -> BatchedInferenceInputs:
...
-                + [0.0] * (max_items_trained_on - max_items_for_this_data_type),
-                dtype=torch.float32,
-            )
-            inputs_mask[data_type].unsqueeze_(0)  # Add batch dimension
-            for _, nc_data in sync_point.data[data_type].items():
-                tensor = DATA_TYPE_TO_BATCHED_NC_DATA_CLASS[data_type].from_nc_data(
-                    nc_data
-                )
+
+            batched_nc_data_class = DATA_TYPE_TO_BATCHED_NC_DATA_CLASS[data_type]
```
Comment (aditya-neuraco) [inline neuracore/ml/utils/policy_inference.py]: Can you add type annotation here?

File: neuracore/ml/utils/policy_inference.py [hunk 2]
```diff
@@ -142,34 +142,42 @@ def _preprocess(self, sync_point: SynchronizedPoint) -> BatchedInferenceInputs:
         # We need to go from sync_point (single time step) to BatchedNCData
         for data_type in sync_point.data.keys():
             inputs[data_type] = []
-            max_items_for_this_data_type = len(sync_point.data[data_type])
+            mask: list[float] = []
+            embodiment_names = self.input_embodiment_description[data_type]
```
Comment (aditya-neuraco) [inline neuracore/ml/utils/policy_inference.py:145]: We should do this for the output embodiment description as well.

File: neuracore/ml/utils/policy_inference.py [hunk 3]
```diff
@@ -142,34 +142,42 @@ def _preprocess(self, sync_point: SynchronizedPoint) -> BatchedInferenceInputs:
...
             if trained_statistics is None:
                 raise ValueError(
                     f"Model was not trained with input statistics for {data_type}."
                 )
             max_items_trained_on = len(trained_statistics)
-            if max_items_for_this_data_type > max_items_trained_on:
+            if len(sync_point.data[data_type]) > max_items_trained_on:
                 raise ValueError(
-                    f"Received {max_items_for_this_data_type} items for data type "
+                    f"Received {len(sync_point.data[data_type])} items for data type "
```
Comment (aditya-neuraco) [inline neuracore/ml/utils/policy_inference.py]: In line 143, you can access this by doing `.items()` on `sync_point.data`. So we can avoid these two function calls. You can call it as `max_items_for_this_data_type` like before. It doesn't seem to be used anywhere else.

File: neuracore/ml/utils/policy_inference.py [hunk 4]
```diff
@@ -142,34 +142,42 @@ def _preprocess(self, sync_point: SynchronizedPoint) -> BatchedInferenceInputs:
...
         for data_type in sync_point.data.keys():
             inputs[data_type] = []
-            max_items_for_this_data_type = len(sync_point.data[data_type])
+            mask: list[float] = []
+            embodiment_names = self.input_embodiment_description[data_type]
             trained_statistics = self.input_dataset_statistics.get(data_type)
             if trained_statistics is None:
                 raise ValueError(
                     f"Model was not trained with input statistics for {data_type}."
                 )
```
Comment (aditya-neuraco) [inline neuracore/ml/utils/policy_inference.py:150]: This should include both input and output stats right?

File: neuracore/ml/utils/policy_inference.py [hunk 5]
```diff
@@ -142,34 +142,42 @@ def _preprocess(self, sync_point: SynchronizedPoint) -> BatchedInferenceInputs:
...
+
                 tensor = apply_preprocessing_methods(
                     batched_data=tensor,
-                    methods=self.input_preprocessing_config.get(data_type, []),
+                    methods=preprocessing_methods,
                 )
                 inputs[data_type].append(tensor)
+
+            inputs_mask[data_type] = torch.tensor(mask, dtype=torch.float32)
+            inputs_mask[data_type].unsqueeze_(0)  # Add batch dimension
```
Comment (aditya-neuraco) [inline neuracore/ml/utils/policy_inference.py]: Nit: IMO, this is a better way to write this function. It's about 7% faster.
- Initialize the `mask` with zeros
- flip the if condition where we append to the mask.
- Chain these two lines: 
```python
inputs_mask[data_type] = torch.tensor(mask, dtype=torch.float32)
inputs_mask[data_type].unsqueeze_(0)  # Add batch dimension
```



```python
  for data_type in sync_point.data.keys():
      inputs[data_type] = []
      embodiment_names = self.input_embodiment_description[data_type]
      trained_statistics = self.input_dataset_statistics.get(data_type)
      if trained_statistics is None:
          raise ValueError(
              f"Model was not trained with input statistics for {data_type}."
          )
      max_items_trained_on = len(trained_statistics)
      if len(sync_point.data[data_type]) > max_items_trained_on:
          raise ValueError(
              f"Received {len(sync_point.data[data_type])} items for data type "
              f"{data_type}, but model was trained on maximum of "
              f"{max_items_trained_on} items."
          )

      mask: list[float] = [0.0] * max_items_trained_on
      batched_nc_data_class = DATA_TYPE_TO_BATCHED_NC_DATA_CLASS[data_type]
      preprocessing_methods = self.input_preprocessing_config.get(data_type, [])

      for index in range(max_items_trained_on):
          name = embodiment_names.get(index)
          if name is None or name not in sync_point.data[data_type]:
              tensor = batched_nc_data_class.sample(batch_size=1, time_steps=1)
          else:
              nc_data = sync_point.data[data_type][name]
              tensor = batched_nc_data_class.from_nc_data(nc_data)
              mask[index] = 1.0

          tensor = apply_preprocessing_methods(
              batched_data=tensor,
              methods=preprocessing_methods,
          )
          inputs[data_type].append(tensor)

      inputs_mask[data_type] = torch.tensor(mask, dtype=torch.float32).unsqueeze(0)
```

File: tests/unit/ml/utils/test_policy_inference.py
Comment (aditya-neuraco) [inline tests/unit/ml/utils/test_policy_inference.py:1]: Add `output_embodiment_description` logic and tests for it as well.

**Review by aditya-neuraco — APPROVED** [review]
LGTM w/ comments.

File: neuracore/ml/utils/policy_inference.py (hunk 3 above)
Comment (ypang-neuraco) [reply neuracore/ml/utils/policy_inference.py]: renamed to max_items_received but we need to keep `sync_point.data[data_type]` because we need to count the items in the data_type, not the number of datatypes in the sync_point

File: neuracore/ml/utils/policy_inference.py (hunk 2 above)
Comment (ypang-neuraco) [reply neuracore/ml/utils/policy_inference.py:145]: Discussed offline, we don't need to pad the output because we can just drop the extra data

File: tests/unit/ml/utils/test_policy_inference.py
Comment (ypang-neuraco) [reply tests/unit/ml/utils/test_policy_inference.py:1]: Discussed offline, we don't need to pad the output because we can just drop the extra data

File: neuracore/ml/utils/policy_inference.py (hunk 4 above)
Comment (ypang-neuraco) [reply neuracore/ml/utils/policy_inference.py:150]: Todo when normalisation move to preprocessing
